### PR TITLE
RBD not attach after node reboot

### PR DESCRIPTION
**Pull request by wenlxie**
After the node reboot, the symbol link for RBD  volumes exists but the device has gone.
Volume manager just check whether the symbol link exists or not , this will make  RBD device not attach after node reboot
```release-notes
Fix the issue that RBD device not get attached after node reboot.
```

### CODE REVIEW

**Comment by wenlxie** [comment]
@mtanino  @sbezverk  PTAL

**Comment by wenlxie** [comment]
Code: https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_executor.go#L920
It just check the symbol link but not the device.

**Comment by mtanino** [comment]
/ok-to-test
@mkimuram PTAL

File: pkg/volume/rbd/rbd.go [hunk 1]
```diff
@@ -401,6 +401,23 @@ func (plugin *rbdPlugin) ConstructBlockVolumeSpec(podUID types.UID, volumeName,
 	pluginDir := plugin.host.GetVolumeDevicePluginDir(rbdPluginName)
 	blkutil := volumepathhandler.NewBlockVolumePathHandler()
 
+	devicePath, err := filepath.EvalSymlinks(mapPath)
+	if err != nil {
+		return nil, fmt.Errorf("Can't get device from the volume :%s map path %s:%v", volumeName, mapPath)
```
Comment (wgliang) [inline pkg/volume/rbd/rbd.go]: Maybe you have a %s more.

Comment (wenlxie) [reply pkg/volume/rbd/rbd.go]: @wgliang  Thanks. It is  changed.

**Comment by mkimuram** [comment]
@wenlxie  

Is my understanding below right? 
- rbd's ConstructBlockVolumeSpec() calls FindGlobalMapPathUUIDFromPod() to get globalMapPathUUID,
- However, FindGlobalMapPathUUIDFromPod() returns err(*1), if actual device symlinked by mapPath doesn't exist,
- As a result, ConstructBlockVolumeSpec returns nil and reconstructVolume() in reconciler fails.

So, this patch is trying to make ConstructBlockVolumeSpec() return volumeSpec without calling FindGlobalMapPathUUIDFromPod(), which makes volumeSpec's v1.RBDPersistentVolumeSource to be {}.

I'll continue to check below and give further feedback:
- If there are another way than to call FindGlobalMapPathUUIDFromPod() there,
- If reconstruction be safely done without PersistentVolumeSource.

Note that:
- Many other plugins call FindGlobalMapPathUUIDFromPod() in ConstructBlockVolumeSpec(), so the same issue will happen to them if device disappears on reboot on these plugins. 
- CheckVolumeExistenceOperation() is called after ReconstructVolumeOperation(), so it seems that adding check for existence of actual device in CheckVolumeExistenceOperation() won't be of help.

(*1) FindGlobalMapPathUUIDFromPod()
   https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/volumepathhandler/volume_path_handler.go#L205

**Comment by wenlxie** [comment]
@mkimuram 
Thanks for the comments.

The issue is after node reboot, the symLink exists but its device has gone.
So FindGlobalMapPathUUIDFromPod() will return the volumeSpec instead of  err(*1)
Then this volume will be put at dsw and asw (https://github.com/kubernetes/kubernetes/blob/release-1.9/pkg/kubelet/volumemanager/reconciler/reconciler.go#L414). The container will start because it seems that the volume has been mounted.

I am porting block device feature to k8s 1.9,  seems this issue just in k8s1.9.

At k8s 1.10, syncStates() has been refactored.
I checked the code, this issue should be fixed at k8s1.10 (Not have env to test, you can have a try if you have time)
At code: 
https://github.com/kubernetes/kubernetes/blob/release-1.10/pkg/kubelet/volumemanager/reconciler/reconciler.go#L390
This volume will not be put at ASW, so I think it will be attached again at reconcile() 

I will close this PR.

**Pull request closed without merging**